### DISCUSSION TITLE
Is there any way I can predict TimesNet without retraining?

### DISCUSSION THREAD

**Discussion by Meekwest** [Nixtla/neuralforecast#1209] [Q&A]
Greetings to all. I can't figure out how to use TimesNet for short-term forecasting after training and whether it is possible to make predictions without retraining. For example, I have historical data from 2018 to 2023, where I train a TimesNet model on a time series. These are indicators of the value of the company's shares in the form of

- 02.01.2024 07:00 297.34
- 02.01.2024 08:00 297.35
- 02.01.2024 09:00 297.36
- 02.01.2024 10:00 297.38
- 02.01.2024 11:00 297.42
- 02.01.2024 12:00 297.34
- 02.01.2024 13:00 297.34
- 02.01.2024 14:00 297.34
- 02.01.2024 15:00 297.31
- 02.01.2024 16:00 298.43

I set the following model parameters during training:
```
horizon = 2  # future 2 hours
model = TimesNet(h=horizon,
                               input_size=5*horizon,
                               max_steps=100,
                               top_k=3,
                               num_kernels=3,
                               batch_size=2,
                               windows_batch_size=32,
                               learning_rate=0.001,
                               futr_exog_list=[],
                               scaler_type='standard')
```

**The last line in my training dataset is 02.01.2024 16:00 298.43**
Immediately after learning, I try to predict on

- 02.01.2024 17:00
- 02.01.2024 18:00

TimesNet is doing well. But then I want to **predict at 19:00 and 20:00**, but already upload fresh data (because at the current time I already know what the actual amounts were at 17:00 and 18:00). I can't put fresh data into the model, I just can't predict on the trained model at 19:00 and 20:00 either, since **it wants to see data for 17:00 and 18:00** in trained dataset. So in order to predict a couple of hours ahead, given the actual data for the last 10 hours, do I need to retrain the model every time?
![image](https://github.com/user-attachments/assets/e6506e26-a2ed-4208-95c7-59d1ba19f078)

**Comment by jmoralez** [comment]
You don't need to retrain the model, you just have to update your data and provide it through the `df` argument, e.g.
```python
nf.fit(train_df)
extended = pd.concat([train_df, updates_df])
nf.predict(df=extended)
```
where `updates_df` contains the values for 19:00 and 20:00.

If you're using exogenous features then you have to provide their values for 21:00, 22:00, ...

**Reply by Meekwest** [reply]
Thank you very much. Really works! One more question, please. How can I reduce the loss metric on each epoch? She jumps from epoch to epoch with me. One can have 1.05 on one, 0.8 on the other, 0.3 on the third, and 0.8 on the fourth again
